ttnn/cpp/ttnn/tensor/tensor_impl.cpp

```diff
@@ -840,6 +847,214 @@ Tensor to_device<bfloat8_b>(
+std::vector<T> convert_logical_data_to_physical_data(
+    const std::vector<T>& logical_data, const TensorSpec& tensor_spec) {
+    const auto& logical_shape = tensor_spec.logical_shape();
+    TT_FATAL(
+        logical_data.size() == logical_shape.volume(),
+        "Logical data size {} should be same as volume indicated by logical shape {}",
+        logical_data.size(),
+        logical_shape);
+
+    const auto& physical_shape = tensor_spec.physical_shape();
+    auto [logical_shard_shape, physical_shard_shape] = get_logical_and_physical_shard_shapes(tensor_spec);
+
+    std::vector<T> physical_data(physical_shape.height() * physical_shape.width(), 0);
+
+    auto logical_2D_shape = flatten_to_2D(logical_shape);
+    size_t physical_stride = physical_shape.width();
+
+    const auto logical_physical_mapping = compute_logical_to_physical_shards_mapping(
+        logical_2D_shape, logical_shard_shape, physical_shard_shape, physical_stride);
+
+    for (const auto& [indices, cols] : logical_physical_mapping) {
+        for (const auto& idx_pair : indices) {
+            auto logical_idx_start = idx_pair[0];
+            auto physical_idx_start = idx_pair[1];
+
+            for (size_t col = 0; col < cols; col++) {
+                physical_data[physical_idx_start + col] = logical_data[logical_idx_start + col];
+            }
+        }
+    }
+
+    TT_FATAL(
+        physical_data.size() == physical_shape.height() * physical_shape.width(),
+        "Physical data size {} should be same as volume indicated by physical shape {}",
+        physical_data.size(),
+        physical_shape);
+
+    if (tensor_spec.layout() == Layout::TILE) {
+        // TODO: Fix convert_layout_row_major_to_tile to take in vector instead of buffer?
+        return tensor_impl::convert_layout_row_major_to_tile(
+            physical_shape, tensor_spec.tile(), owned_buffer::create(std::move(physical_data)));
+    }
+    return physical_data;
+}
+
+template std::vector<bfloat16> convert_logical_data_to_physical_data<bfloat16>(
+    const std::vector<bfloat16>& logical_data, const TensorSpec& tensor_spec);
+template std::vector<float> convert_logical_data_to_physical_data<float>(
+    const std::vector<float>& logical_data, const TensorSpec& tensor_spec);
+template std::vector<int32_t> convert_logical_data_to_physical_data<int32_t>(
+    const std::vector<int32_t>& logical_data, const TensorSpec& tensor_spec);
+template std::vector<uint32_t> convert_logical_data_to_physical_data<uint32_t>(
+    const std::vector<uint32_t>& logical_data, const TensorSpec& tensor_spec);
+template std::vector<uint16_t> convert_logical_data_to_physical_data<uint16_t>(
+    const std::vector<uint16_t>& logical_data, const TensorSpec& tensor_spec);
+template std::vector<uint8_t> convert_logical_data_to_physical_data<uint8_t>(
+    const std::vector<uint8_t>& logical_data, const TensorSpec& tensor_spec);
+
+template <typename T>
+std::vector<T> convert_physical_data_to_logical_data(
```

Comment:
What is your thinking about why this should be an external, publicly available method (symbol available to library consumers) versus being a detail on implementation and a method/constructor in Tensor?

Reply:
`tensor_impl` entire namespace I thought was more internal implementation. Should probably reorganize everything in that namespace.